File: cellulosesz-api/src/main/java/top/likoslupus/cellulosesz/api/permission/PermissionService.java
package top.likoslupus.cellulosesz.api.permission;

import top.likoslupus.cellulosesz.api.platform.CellPlayer;

import java.util.Optional;

/**
 * Permission queries for a stable player identity. Backend failures are propagated, not represented
 * as absence.
 */
public interface PermissionService {

    boolean has(
            CellPlayer player,
            String permission
    );

    int intOption(
            CellPlayer player,
            String key,
            int fallback
    );

    boolean boolOption(
            CellPlayer player,
            String key,
            boolean fallback
    );

    Optional<String> stringOption(
            CellPlayer player,
            String key
    );

}

File: cellulosesz-api/src/main/kotlin/top/likoslupus/cellulosesz/api/permission/PermissionService.kt
package top.likoslupus.cellulosesz.api.permission

import top.likoslupus.cellulosesz.api.platform.CellPlayer

/**
 * Permission queries for a stable player identity. Backend failures are propagated, not represented
 * as absence.
 */
public interface PermissionService {

    public fun has(
        player: CellPlayer,
        permission: String
    ): Boolean

    public fun intOption(
        player: CellPlayer,
        key: String,
        fallback: Int
    ): Int

    public fun boolOption(
        player: CellPlayer,
        key: String,
        fallback: Boolean
    ): Boolean

    public fun stringOption(
        player: CellPlayer,
        key: String
    ): String?

}
